### ISSUE TITLE
Questions about the scope of this project

### ISSUE THREAD

**Issue by tonycpsu** [wamonite/packermate#17]
Hi, I discovered this project while searching for something to make the task of writing and maintaining Packer configs easier, as I find them rather unwieldy to work with.  I was going to do some work trying to adapt it to my particular needs, but I wanted to check first to see how many of these features you would find interesting, or would be willing to accept as contributions in order to expand packermate's capabilities.

Things I'm looking to add would be:
- Support for CentOS as a guest OS -- there seems to be some coupling between the packermate code and Ubuntu as a guest OS, e.g. the preseed stuff in the virtualbox module.  I would like to set up CentOS and Ubuntu as subclasses of a Guest OS base class, which would provide a method (to be overwritten by the guest OSes) for any sort of kickstart/preseed logic.
- Support for ansible provisioning.  I'm not sure if anything in packermate precludes this, but my goal is to have as little logic as possible in the kickstart / shell script provisioning layers, so that most of the application configuration is done in Ansible, in the hopes of only having to write it once, regardless of the base platform
- Support for VMWare as a target
- Support for Docker containers as a target

I've only briefly built the example VM and poked around the code a bit, so I apologize if any of my impressions about packermate's current capabilities are wrong.

Really just wanted to start a conversation to see if any of this sounds useful, if you have plans to build any of these things into it, etc.  And also to just say thanks for putting this out there.

**Comment by wamonite** [comment]
Hi,

Thanks for your interest in packermate! I’ve often considered adding other OSs/targets so would be very interested in contributions. Have not had time to investigate as I’ve used packermate as the foundation of a tool for @spatialbuzz (imagemate) to build, manage, version and launch clusters of VirtualBox and AWS Vagrant images. Using this in production and hoping to move some features in to packermate if I can decide on how to generalise them (don’t want my decisions on versioning/publishing to be too prescriptive). Ubuntu only so far, as all I needed. imagemate bootstraps Ansible for image provisioning, so that’s one candidate for importing. Had considered it a bit overkill for Docker but it the workflow would fit and happy to discuss. Have also hit some limitations and other use cases of the config object, so am in the (slow) process of improving it and separating it out (configmate on GitHub).

Thanks again and look forward to chatting further,

Warren

**Comment by tonycpsu** [comment]
Hi, it's good to hear you're open to contributions.

I am a bit curious what you mean by "overkill" with respect to Ansible.  If it's that you want to keep the size of the docker containers down, the Ansible pieces could be removed once provisioning is complete.  I just see the agentless setup of Ansible as being able to realize my dream of writing application configuration code just once for all build targets, save for a few parameters that would have to be provided as variables by the Packer build for each target.  Maybe this is a pipe dream, but I've grown tired of maintaining steps in Dockerfiles, shell scripts, Puppet modules, etc.

My first task is going to be seeing what it will take to get Packermate to provision a vanilla CentOS machine with Kickstart, etc.  I can put the Ansible bit on hold if you think that's something you'd want to backport from your other project.  Docker and VM target support isn't strictly necessary for me right now, but would be nice to haves so I can retire my existing projects for building out those platforms.

My initial plan is just to have a class for describing the bits of config that vary (kickstart/preseed logic), then implement subclasses for Ubuntu and Redhat.  If you have any guidance on how you think I should factor out guest OS support / any gotchas to watch out for, let me know.  

Thanks again.
-Tony

**Comment by wamonite** [comment]
Hi Tony,

Sorry, work has been hectic!

I no longer use Ansible within Dockerfiles personally but am happy for them to be supported as build targets.

Good luck with the kickstart/preseed refactoring!

Warren
